### PR TITLE
Optimism v2

### PR DESCRIPTION
**Pull request by brunobar79**
Working fine on optimism mainnet. PoW:  https://recordit.co/fbB0WpmPn0

Fixes RNBW-1594

### CODE REVIEW

File: ios/Podfile.lock [hunk 1]
```diff
@@ -1057,4 +1057,4 @@ SPEC CHECKSUMS:
 
 PODFILE CHECKSUM: 341e5b87722a11ae3c39865fdd0d2673fa60fea9
 
-COCOAPODS: 1.11.0
+COCOAPODS: 1.10.2
```
Comment (jinchung) [inline ios/Podfile.lock:1060]: can we leave this change out

File: src/parsers/gas.js [hunk 2]
```diff
@@ -119,19 +121,40 @@ export const defaultGasPriceFormat = (option, timeWait, value) => {
  * @param {Object} prices
  * @param {Number} gasLimit
  */
-export const parseTxFees = (gasPrices, priceUnit, gasLimit, nativeCurrency) => {
+export const parseTxFees = (
+  gasPrices,
+  priceUnit,
+  gasLimit,
+  nativeCurrency,
+  l1GasFeeOptimism
```
Comment (jinchung) [inline src/parsers/gas.js:129]: 🤮

File: src/parsers/gas.js [hunk 3]
```diff
@@ -119,19 +121,40 @@ export const defaultGasPriceFormat = (option, timeWait, value) => {
...
-  const amount = multiply(gasPrice, gasLimit);
+const getTxFee = (
+  gasPrice,
+  gasLimit,
+  priceUnit,
+  nativeCurrency,
+  l1GasFeeOptimism
+) => {
+  let amount = multiply(gasPrice, gasLimit);
+  if (l1GasFeeOptimism && greaterThan(l1GasFeeOptimism, '0')) {
```
Comment (jinchung) [inline src/parsers/gas.js:155]: you need `greaterThan` from `react-native-reanimated` ? the one in `utilities` isn't good enough for u?

File: src/redux/data.js [hunk 4]
```diff
@@ -529,10 +529,10 @@ export const addressAssetsReceived = (
     const restOfTheAssets = existingAssets.filter(
       asset => asset.network !== assetsNetwork
     );
-
+    parsedAssets = concat(parsedAssets, restOfTheAssets);
     parsedAssets = uniqBy(
-      concat(parsedAssets, restOfTheAssets),
-      item => item.uniqueId
+      parsedAssets,
+      item => `${item.uniqueId}_${item.network}`
```
[the remark this answers is not in the document]
Comment (jinchung) [reply src/redux/data.js:535]: hmm gonna think about this

[the remark this answers is not in the document]
Comment (jinchung) [reply src/redux/data.js:535]: also would we need similar behavior in line 546?

[the remark this answers is not in the document]
Comment (jinchung) [reply src/redux/data.js:535]: line of thinking: uniqueId is not actually a uniqueId anymore, and could cause unexpected behavior in other parts of the app (e.g, just thinking out loud: in recycler list view potentially, what about send sheet updates when a user selects arb ETH and then switches to optimism ETH, does it behave as expected, etc)

[the remark this answers is not in the document]
Comment (jinchung) [reply src/redux/data.js:535]: there are a few other places of concern, we should test these out

File: src/redux/gas.js [hunk 5]
```diff
@@ -138,12 +137,16 @@ export const gasPricesStartPolling = (network = networkTypes.mainnet) => async (
   };
 
   const getOptimismGasPrices = async () => {
+    const provider = await getProviderForNetwork(networkTypes.optimism);
+    const baseGasPrice = await provider.getGasPrice();
+    const gasPriceGwei = weiToGwei(baseGasPrice.toString());
```
Comment (jinchung) [inline src/redux/gas.js:142]: can just convert to Number here instead of in line 145, 147, 149

File: src/screens/SendSheet.js [hunk 6]
```diff
@@ -464,7 +465,31 @@ export default function SendSheet(props) {
         );
 
         if (!lessThan(updatedGasLimit, gasLimit)) {
-          updateTxFee(updatedGasLimit, null, currentNetwork);
+          if (network === networkTypes.optimism) {
```
Comment (jinchung) [inline src/screens/SendSheet.js:468]: can we create a separate function for this if else statement and reuse below in like 777

File: src/handlers/web3.js [hunk 7]
```diff
@@ -433,23 +433,25 @@ export const getDataForNftTransfer = (from, to, asset) => {
...
  * @return {String}
  */
-export const estimateGasLimit = async (
-  { asset, address, recipient, amount },
-  addPadding = false,
-  provider = null,
-  network = NetworkTypes.mainnet
+
+export const buildTransaction = async (
+  { asset, address, recipient, amount, gasLimit },
```
Comment (jinchung) [inline src/handlers/web3.js:444]: abc

File: src/handlers/web3.js [hunk 8]
```diff
@@ -458,20 +460,39 @@ export const estimateGasLimit = async (
...
+  return { ...txData, gasLimit };
+};
+
+/**
+ * @desc estimate gas limit
+ * @param {Object} [{selected, address, recipient, amount, gasPrice}]
+ * @return {String}
+ */
+export const estimateGasLimit = async (
+  { asset, address, recipient, amount },
```
Comment (jinchung) [inline src/handlers/web3.js:486]: abc

File: src/utils/ethereumUtils.js [hunk 9]
```diff
@@ -515,7 +524,35 @@ async function parseEthereumUrl(data) {
   });
 }
 
+const calculateL1FeeOptimism = async (tx, provider) => {
+  try {
+    tx.value = toHex(tx.value);
+    tx.nonce = Number(await provider.getTransactionCount(tx.from));
+    delete tx.from;
+    tx.to = toChecksumAddress(tx.to);
```
Comment (jinchung) [inline src/utils/ethereumUtils.js:532]: does this need to be checksummed?

File: src/utils/ethereumUtils.js [hunk 10]
```diff
@@ -515,7 +524,35 @@ async function parseEthereumUrl(data) {
...
+
+    const OVM_GasPriceOracle = new Contract(
+      OVM_GAS_PRICE_ORACLE,
+      optimismGasOracleAbi,
+      provider
+    );
+    const l1FeeInWei = await OVM_GasPriceOracle.getL1Fee(serializedTx);
+    return l1FeeInWei;
+  } catch (e) {
+    logger.log('error calculating l1 fee', e);
```
Comment (jinchung) [inline src/utils/ethereumUtils.js:550]: no return value for when we fail to get the l1 fee

File: src/utils/ethereumUtils.js [hunk 11]
```diff
@@ -515,7 +524,35 @@ async function parseEthereumUrl(data) {
   });
 }
 
+const calculateL1FeeOptimism = async (tx, provider) => {
+  try {
+    tx.value = toHex(tx.value);
+    tx.nonce = Number(await provider.getTransactionCount(tx.from));
```
Comment (jinchung) [inline src/utils/ethereumUtils.js:530]: hmmm

File: src/screens/SendConfirmationSheet.js [hunk 12]
```diff
@@ -272,6 +273,7 @@ export default function SendConfirmationSheet() {
 
   const shouldShowChecks =
     isL2 &&
+    network !== networkTypes.optimism &&
```
Comment (jinchung) [inline src/screens/SendConfirmationSheet.js:276]: note to revert this change

File: src/utils/ethereumUtils.js (hunk 11 above)
Comment (jinchung) [reply src/utils/ethereumUtils.js:530]: will this be affected by chris' nonce mgmt stuff?

File: src/handlers/web3.js (hunk 7 above)
Comment (brunobar79) [reply src/handlers/web3.js:444]: wanna turn on an eslint rule for this so we can forget about it?

File: src/redux/data.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (brunobar79) [reply src/redux/data.js:535]: `uniqueId is not actually a unique ID anymore` - Can you point to a specific place?
IMO I am making this even more unique (not less), so it shouldn't bring additional problems as long as it's not treated as an address anywhere (it shouldn't). 

Can you please list all the other places of concern so we add it as QA notes?

File: src/parsers/gas.js (hunk 2 above)
Comment (brunobar79) [reply src/parsers/gas.js:129]: Do you have a suggestion?

File: src/utils/ethereumUtils.js (hunk 11 above)
Comment (brunobar79) [reply src/utils/ethereumUtils.js:530]: this is not a problem in optimism because txs are confirmed way faster than you can complete any flow.
However, after #2532 is merged I can get the nonce through the nonce manager instead

File: src/utils/ethereumUtils.js (hunk 9 above)
Comment (brunobar79) [reply src/utils/ethereumUtils.js:532]: yes, otherwise it was failing

File: src/utils/ethereumUtils.js (hunk 10 above)
Comment (brunobar79) [reply src/utils/ethereumUtils.js:550]: that was intended.  If it's undefined it will default to null and won't be added to the L2 gas fee.

File: src/screens/SendSheet.js (hunk 6 above)
Comment (brunobar79) [reply src/screens/SendSheet.js:469]: done

File: src/redux/data.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (jinchung) [reply src/redux/data.js:535]: that's what i mean - uniqueId is used in other places in the app with the expectation that it is actually unique (bc it just uses address most of the time) - but now it is no longer the case

File: src/parsers/gas.js (hunk 2 above)
Comment (jinchung) [reply src/parsers/gas.js:129]: just complaining about how we're forced to do this change :)

**Review by skylarbarrera — APPROVED** [review]
🙆‍♂️

**Review by jinchung — APPROVED** [review]
Fixes to address uniqueId issue will be in a separate PR

**Review by skylarbarrera — APPROVED** [review]
last few bugs are fixed, send + wc working as expected :shipit:
